public/app/features/alerting/unified/components/notification-policies/useNotificationPolicyRoute.ts

```diff
@@ -0,0 +1,162 @@
+  ComGithubGrafanaGrafanaPkgApisAlertingNotificationsV0Alpha1RoutingTreeSpec,
+  generatedRoutesApi,
+} from '../../openapi/routesApi.gen';
+import { K8sAnnotations, PROVENANCE_NONE, ROOT_ROUTE_NAME } from '../../utils/k8s/constants';
+import { getAnnotation, getK8sNamespace, shouldUseK8sApi } from '../../utils/k8s/utils';
+const k8sRoutesToRoutesMemoized = memoize(k8sRoutesToRoutes, { maxSize: 1 });
+
+const { useListNamespacedRoutingTreeQuery, useReplaceNamespacedRoutingTreeMutation } = generatedRoutesApi;
+
+const {
+  useUpdateAlertmanagerConfigurationMutation,
+  useLazyGetAlertmanagerConfigurationQuery,
+  useGetAlertmanagerConfigurationQuery,
+} = alertmanagerApi;
+
+export const useNotificationPolicyRoute = ({ alertmanager }: BaseAlertmanagerArgs, { skip }: Skippable = {}) => {
+  const k8sApiSupported = shouldUseK8sApi(alertmanager);
+
+  const k8sRouteQuery = useListNamespacedRoutingTreeQuery(
+    { namespace: getK8sNamespace() },
+    {
+      skip: skip || !k8sApiSupported,
+      selectFromResult: (result) => {
+        return {
+          ...result,
+          currentData: result.currentData ? k8sRoutesToRoutesMemoized(result.currentData.items) : undefined,
+          data: result.data ? k8sRoutesToRoutes(result.data.items) : undefined,
+        };
+      },
+    }
+  );
+
+  const amConfigQuery = useGetAlertmanagerConfigurationQuery(alertmanager, {
+    skip: skip || k8sApiSupported,
+    selectFromResult: (result) => {
+      return {
+        ...result,
+        currentData: result.currentData?.alertmanager_config?.route
+          ? [result.currentData.alertmanager_config.route]
+          : undefined,
+        data: result.data?.alertmanager_config?.route ? result.data.alertmanager_config.route : undefined,
+      };
+    },
+  });
+
+  return k8sApiSupported ? k8sRouteQuery : amConfigQuery;
+};
+
+export function useUpdateNotificationPolicyRoute(selectedAlertmanager: string) {
+  const [getAlertmanagerConfiguration] = useLazyGetAlertmanagerConfigurationQuery();
+  const [updateAlertmanagerConfiguration] = useUpdateAlertmanagerConfigurationMutation();
+
+  const [updatedNamespacedRoute] = useReplaceNamespacedRoutingTreeMutation();
+
+  const k8sApiSupported = shouldUseK8sApi(selectedAlertmanager);
+
+  async function updateUsingK8sApi({ newRoute }: { newRoute: Route }) {
+    const namespace = getK8sNamespace();
+    const { routes, ...rest } = newRoute;
+    const { provenance, ...defaults } = rest;
+    // Convert Route to K8s compatible format
+    const k8sRoute: ComGithubGrafanaGrafanaPkgApisAlertingNotificationsV0Alpha1RoutingTreeSpec = {
+      defaults: {
+        ...defaults,
+        // TODO: Fix types in k8s API? Fix our types to not allow empty receiver? TBC
+        receiver: defaults.receiver || '',
+      },
+      routes: newRoute.routes?.map(routeToK8sSubRoute) || [],
+    };
+
+    // TODO Add a check to see if not updated in the meantime
```

Comment:
(will add this in, just need to send resourceVersion along with the request. There's a bug to be fixed on the backend where changing matchers doesn't update resourceVersion)